### PR TITLE
Experimental - oneOf directive support

### PR DESCRIPTION
**Pull request by bbakerman**
This PR provides `@oneOf` support to graphql-java

related to https://github.com/graphql-java/graphql-java/issues/3266

This is now complete.  

The error messages here are inspired by the graphql-js versions.

I chose to add `isOneOf` into our pre built Introspection queries  by default and have an option to turn it off if it causes problems (I dont see how it would be just in case)

I am a fan of the directive based approach of this mechanism

This code is considered experimental namely because the spec has not been accepted.  The PR has been put into the reference implementation graphql-js however so its not pie in the sky stuff

### CODE REVIEW

File: src/main/java/graphql/introspection/IntrospectionQueryBuilder.java [hunk 1]
```diff
@@ -85,6 +92,7 @@ public static Options defaultOptions() {
                     true,
                     false,
                     true,
+                    true,
```
Comment (bbakerman) [inline src/main/java/graphql/introspection/IntrospectionQueryBuilder.java:95]: on by default in new versions

Comment (bbakerman) [reply src/main/java/graphql/introspection/IntrospectionQueryBuilder.java:95]: but some one could turn it off if need be

File: src/test/groovy/graphql/execution/ValuesResolverTest.groovy [hunk 2]
```diff
@@ -345,6 +346,162 @@ class ValuesResolverTest extends Specification {
...
+
+        '`$var` { var: { a: "abc" } }' | VariableReference.of("var")
+                                                      | CoercedVariables.of([var: [a: "abc"]]) | [arg: [a: "abc"]]
+
+        '{ a: $var }` { var : "abc"}'  | buildObjectLiteral([
+                a: VariableReference.of("var")
+        ])                                            | CoercedVariables.of([var: "abc"])      | [arg: [a: "abc"]]
+
+    }
+
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/ValuesResolverTest.groovy:504]: most of the tests are here - these are taken from the examples given in the spec update plus some they missed

File: src/test/groovy/graphql/schema/GraphQLInputObjectTypeTest.groovy [hunk 3]
```diff
@@ -79,4 +83,101 @@ class GraphQLInputObjectTypeTest extends Specification {
...
+
+        when:
+        ei = ExecutionInput.newExecutionInput('query q($var : OneOf)  { f( arg : $var) { key value }}').variables([var: [a: null]]).build()
+        er = graphQL.execute(ei)
+        then:
+        !er.errors.isEmpty()
+        er.errors[0].message == "Exception while fetching data (/f) : OneOf type field 'OneOf.a' must be non-null."
+
+        // lots more covered in unit tests
+    }
```
Comment (bbakerman) [inline src/test/groovy/graphql/schema/GraphQLInputObjectTypeTest.groovy:182]: some end 2 end testing behind the unit tested we have for ValueResolver

File: src/main/java/graphql/execution/ValuesResolver.java [hunk 4]
```diff
@@ -373,12 +375,42 @@ private static Map<String, Object> getArgumentValuesImpl(
...
+            throw new OneOfTooManyKeysException(msg);
+        }
+        String fieldName = objectMap.keySet().iterator().next();
+        if (objectMap.get(fieldName) == null) {
+            String msg = I18n.i18n(I18n.BundleType.Execution, locale)
+                    .msg("Execution.handleOneOfValueIsNullError", oneOfInputType.getName() + "." + fieldName);
+            throw new OneOfNullValueException(msg);
+        }
+    }
+
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ValuesResolver.java:413]: remember we can have a mix of input - literals and maps and invalid input might be a with a variable with no map entries or multiple.

The unit tests for this explore these different conditions

File: src/main/java/graphql/schema/GraphQLInputObjectType.java [hunk 5]
```diff
@@ -60,18 +62,37 @@ private GraphQLInputObjectType(String name,
...
                 (fld1, fld2) -> assertShouldNeverHappen("Duplicated definition for field '%s' in type '%s'", fld1.getName(), this.name)));
     }
 
+    private boolean hasOneOf(List<GraphQLDirective> directives, List<GraphQLAppliedDirective> appliedDirectives) {
+        if (appliedDirectives.stream().anyMatch(d -> Directives.OneOfDirective.getName().equals(d.getName()))) {
+            return true;
+        }
+        // eventually GraphQLDirective as applied directive goes away
+        return directives.stream().anyMatch(d -> Directives.OneOfDirective.getName().equals(d.getName()));
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/schema/GraphQLInputObjectType.java:79]: The boolean is calculated on creation and hence fast from that time forward

File: src/main/java/graphql/schema/GraphQLSchema.java [hunk 6]
```diff
@@ -917,6 +917,9 @@ private GraphQLSchema buildImpl() {
             if (additionalDirectives.stream().noneMatch(d -> d.getName().equals(Directives.SpecifiedByDirective.getName()))) {
                 additionalDirectives.add(Directives.SpecifiedByDirective);
             }
+            if (additionalDirectives.stream().noneMatch(d -> d.getName().equals(Directives.OneOfDirective.getName()))) {
+                additionalDirectives.add(Directives.OneOfDirective);
+            }
```
Comment (bbakerman) [inline src/main/java/graphql/schema/GraphQLSchema.java:922]: always available like deprecated and specifiedby

**Review by dondonz — APPROVED** [review]
Woo hoo!

**Pull request merged**